metricbeat/module/system/diskio/diskstat_windows_helper.go

```diff
@@ -0,0 +1,232 @@
+	Name    string
+	UNCPath string
+}
+
+type diskPerformance struct {
+	BytesRead           int64
+	BytesWritten        int64
+	ReadTime            int64
+	WriteTime           int64
+	IdleTime            int64
+	ReadCount           uint32
+	WriteCount          uint32
+	QueueDepth          uint32
+	SplitCount          uint32
+	QueryTime           int64
+	StorageDeviceNumber uint32
+	StorageManagerName  [8]uint16
+}
+
+// ioCounters gets the diskio counters and maps them to the list of counterstat objects.
+func ioCounters(names ...string) (map[string]disk.IOCountersStat, error) {
+	if err := enablePerformanceCounters(); err != nil {
+		return nil, err
+	}
+	logicalDisks, err := getLogicalDriveStrings()
+	if err != nil || len(logicalDisks) == 0 {
+		return nil, err
+	}
+	ret := make(map[string]disk.IOCountersStat)
+	for _, drive := range logicalDisks {
+		// not get _Total or Harddrive
+		if len(drive.Name) > 3 {
+			continue
+		}
+		// filter by included devices
+		if len(names) > 0 && !containsDrive(names, drive.Name) {
+			continue
+		}
+
+		counter, err := ioCounter(drive.UNCPath)
+		if err != nil {
+			return nil, err
+		}
+		ret[drive.Name] = disk.IOCountersStat{
+			Name:       drive.Name,
+			ReadCount:  uint64(counter.ReadCount),
+			WriteCount: uint64(counter.WriteCount),
+			ReadBytes:  uint64(counter.BytesRead),
+			WriteBytes: uint64(counter.BytesWritten),
+			ReadTime:   uint64(counter.ReadTime),
+			WriteTime:  uint64(counter.WriteTime),
+		}
+	}
+	return ret, nil
+}
+
+// i0Counter calls syscal func CreateFile to generate a handler then executes the DeviceIoControl func in order to retrieve the metrics.
+func ioCounter(path string) (diskPerformance, error) {
+	var diskPerformance diskPerformance
+	var diskPerformanceSize uint32
+	utfPath, err := syscall.UTF16PtrFromString(path)
+	if err != nil {
+		return diskPerformance, err
+	}
+	hFile, err := syscall.CreateFile(utfPath,
+		syscall.GENERIC_READ|syscall.GENERIC_WRITE,
+		syscall.FILE_SHARE_READ|syscall.FILE_SHARE_WRITE,
+		nil,
+		syscall.OPEN_EXISTING,
+		syscall.FILE_FLAG_BACKUP_SEMANTICS,
+		0)
+
+	if err != nil {
+		return diskPerformance, err
+	}
+	defer syscall.CloseHandle(hFile)
+
+	err = syscall.DeviceIoControl(hFile,
+		ioctlDiskPerformance,
+		nil,
+		0,
+		(*byte)(unsafe.Pointer(&diskPerformance)),
+		uint32(unsafe.Sizeof(diskPerformance)),
+		&diskPerformanceSize,
+		nil)
+	if err != nil {
+		return diskPerformance, err
+	}
+	return diskPerformance, nil
+}
+
+// enablePerformanceCounters will enable performance counters by adding the EnableCounterForIoctl registry key
+func enablePerformanceCounters() error {
+	key, err := registry.OpenKey(registry.LOCAL_MACHINE, "SYSTEM\\CurrentControlSet\\Services\\partmgr", registry.READ|registry.WRITE)
+	if err != nil {
+		return errors.Errorf("cannot open new key in the registry in order to enable the performance counters: %s", err)
+	}
+	if err = key.SetDWordValue("EnableCounterForIoctl", 1); err != nil {
```

Comment:
Do you think it would be useful to have some logging telling the user that Metricbeat is modifying the registry value and why?